python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -236,44 +237,49 @@ def no_messages_debug_text(self) -> str:
 WAIT_FOR_LOGS_TIMEOUT = 60
 
 
-class PipesBlobStoreMessageReader(PipesMessageReader):
-    """Message reader that reads a sequence of message chunks written by an external process into a
-    blob store such as S3, Azure blob storage, or GCS.
-
-    The reader maintains a counter, starting at 1, that is synchronized with a message writer in
-    some pipes process. The reader starts a thread that periodically attempts to read a chunk
-    indexed by the counter at some location expected to be written by the pipes process. The chunk
-    should be a file with each line corresponding to a JSON-encoded pipes message. When a chunk is
-    successfully read, the messages are processed and the counter is incremented. The
-    :py:class:`PipesBlobStoreMessageWriter` on the other end is expected to similarly increment a
-    counter (starting from 1) on successful write, keeping counters on the read and write end in
-    sync.
+T_C = TypeVar("T_C")
 
-    If `log_readers` is passed, the message reader will start the passed log readers when the
-    `opened` message is received from the external process.
-
-    Args:
-        interval (float): interval in seconds between attempts to download a chunk
-        log_readers (Optional[Sequence[PipesLogReader]]): A set of readers for logs.
-    """
 
+class PipesThreadedMessageReader(PipesMessageReader):
     interval: float
-    counter: int
-    log_readers: Sequence["PipesLogReader"]
+    log_readers: Mapping[str, "PipesLogReader"]
     opened_payload: Optional[PipesOpenedData]
 
     def __init__(
         self,
         interval: float = 10,
-        log_readers: Optional[Sequence["PipesLogReader"]] = None,
+        log_readers: Optional[
+            Union[Mapping[str, "PipesLogReader"], Sequence["PipesLogReader"]]
+        ] = None,
     ):
         self.interval = interval
-        self.counter = 1
-        self.log_readers = check.opt_sequence_param(
-            log_readers, "log_readers", of_type=PipesLogReader
-        )
+
         self.opened_payload = None
 
+        if isinstance(log_readers, abc.Mapping):
+            self.log_readers = check.dict_param(
+                log_readers, "log_readers", key_type=str, value_type=PipesLogReader
+            )
+        elif isinstance(log_readers, abc.Sequence):
+            self.log_readers = {
+                str(i): lr
+                for i, lr in enumerate(
+                    check.opt_sequence_param(
+                        cast(Sequence["PipesLogReader"], log_readers),
+                        "log_readers",
+                        of_type=PipesLogReader,
+                    )
+                )
+            }
+        else:
+            self.log_readers = {}
+
+        self.runtime_params_updated = Event()
+        self.state = {"params": {"message_reader": {}, "log_readers": {}}}
+
+    @abstractmethod
+    def target_is_readable(self, params: PipesParams) -> bool: ...
```

Comment:
This is worth of a doc string. What does "readable" mean in this context? When is something not readable?

Reply:
Renamed to `can_read`. 

`can_read` should return `False` if: 
- `MessageReader` is missing some params which are expected to be submitted once known with `.update_params`
- perahs the location to read from is not physically available yet